design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,210 @@
+  [docker]
+  auto_discover=true
+
+  [[docker.include]]
+  pattern = "nginx"
+
+  [[docker.exclude]]
+  pattern = "kube-*"
+  ```
+- The `auto_discover` and `include`/`exclude` configs are used by the `tedge-agent`
+  to process the supported types listed by that plugin.
+  Any plugin-specific settings can also be defined in these files, which are ignored by the agent,
+  but can be used by the plugin.
+- The main config file for a plugin must be named after the plugin itself (e,g: `docker.sh` -> `docker.toml`).
+  Key configs like `auto_discover` can only be defined in this main file.
+  The inclusion/exclusion list can be defined in extension files as well,
+  but the table names with the plugin name prefix (e.g: `[docker.include]` and `[docker.exclude]`) must be used in those as well.
+- When new software is installed, their corresponding `include/exclude` entries can be appended to the main plugin config itself,
+  or created in an independent extension file.
+
+The agent uses the plugins as follows:
+
+- On startup, gather the supported types from all the plugins by running the `list` command on them.
+  Those failing to execute the `list` command are not qualified as valid plugins and ignored.
+- Whenever the `/etc/tedge/log-plugins` directory is updated (new plugin installed or existing one removed),
+  agent refreshes the supported types.
+  Simply touching this directory would also trigger a refresh.
+- When the supported types are published over mqtt, their source plugin information is also appended to that type
+  in the format <log_type>::<plugin_type>.
+  For example, if a `journals` plugin lists two different log types: `mosquitto` and `tedge-agent`,
+  both types would be reported as `mosquitto::journald` and `tedge-agent::journald`.
+- When a `log_upload` request is received for a type, call the `get` command of the corresponding plugin for that type
+  which can be derived from the `::` suffix of that type.
+  The `type` passed to the `get` command would not include the plugin suffix.
+  If there is no explicit suffix, it is delegated to the default file plugin.
+
+## Config management
+
+### Phase 1
+
+- Similar to log management, config plugins are defined under `/etc/tedge/config-plugins`.
+- Config plugins can be used to perform any pre/post processing steps before/after a configuration file is updated by the agent.
+- A config plugin needs to support the following sub-commands:
+  - `list`: List all the config types supported by this plugin.
+    Used to detect if this is a valid plugin if it exits with exit code 0.
+    The types must be printed between `:::begin-tedge:::` and `:::end-tedge:::` marker lines,
+    with one type per line.
+  - `get <type> <tmp-target-file-path>`: Get the existing configuration for the given type.
+  - `set <type> [--url <new-config-url>] [--file <new-config-file-path>]`: Update the existing config file
+  - `finalize <type>`: Take any post-processing actions after the new config is applied
```

Comment:
Why a `finalize` action and no `prepare` action?

I guess that one reason is that preparation (e.g. a backup of the existing configuration) can be done as the first step of the `set` command. But then, finalization (e.g. restarting a service) can also be done by the `set` command.

~~=> I'm in favor of no `finalize` command nor `prepare` command.~~

__Oh I see!__ The protocol defined bellow using `get`, `set` and `finalize` to handle commit or rollback is neat.

The issue is more about the doc structure. That's okay in this design doc. However in a user doc, it would be good to tell before hand that the agent triggers updates and handles commits or rollbacks, while the plugin defines how to get, set and use the configuration file.

Reply:
> The issue is more about the doc structure. That's okay in this design doc. However in a user doc, it would be good to tell before hand that the agent triggers updates and handles commits or rollbacks, while the plugin defines how to get, set and use the configuration file

Sure. I guess naming it `finalise` (to keep it consistent with sm plugin) was the mistake. Wondering if naming it `apply` or `commit` would have been clearer.